lib/clib/CCorresLemmas.thy

```diff
@@ -1040,119 +1040,211 @@ lemma ccorres_Guard_True_Seq:
-                             \<and> (s', st) \<in> srel \<and> rv = xf st)"
-    apply (erule exec_While_final_inv''; simp)
+                             \<and> (s', st) \<in> srel \<and> rrel rv (xf st))"
+    apply (erule_tac C="{s'. cond_xf s' \<noteq> 0}" in exec_While_final_inv''; simp)
      apply (fastforce simp: whileLoop_cond_fail return_def)
     apply clarsimp
-    apply (rename_tac t t' t'' s)
-    apply (frule intermediate_Normal_state[where P="G' \<inter> C'"])
+    apply (rename_tac t t' t'' r s y)
+    apply (frule_tac P="{s'. s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')
+                             \<and> s' \<in> C' \<and> (C r s \<noteq> None) \<and> the (C r s)}"
+                  in intermediate_Normal_state)
       apply fastforce
-     apply (fastforce intro: body_inv)
+     apply (fastforce intro: body_inv conseqPre)
     apply clarsimp
     apply (rename_tac inter_t)
-    apply (prop_tac "\<exists>s'. (xf inter_t, s') \<in> fst (B (xf t) s) \<and> (s', inter_t) \<in> srel")
-     subgoal by (erule ccorresE[OF body_ccorres])
-                (fastforce simp: no_fail_def nf[simplified no_fail_def] dest: EHOther)+
+    apply (prop_tac "\<exists>rv' s'. rrel rv' (xf inter_t) \<and> (rv', s') \<in> fst (B r s)
+                              \<and> (s', inter_t) \<in> srel")
+     apply (insert body_ccorres)[1]
+     apply (drule_tac x=r in meta_spec)
+     apply (erule (1) ccorresE)
+         apply fastforce
+        apply fastforce
+       using nf apply (fastforce simp: no_fail_def)
+      apply (fastforce dest!: EHOther)
+     apply fastforce
     apply clarsimp
-    apply (prop_tac "G s'")
+    apply (prop_tac "G rv' s'")
      apply (fastforce dest: use_valid intro: body_inv)
     apply (prop_tac "inter_t \<in> G'")
      apply (fastforce dest: hoarep_exec[rotated] intro: body_inv)
+    apply (drule_tac x=rv' in spec)
     apply (drule_tac x=s' in spec)
+    apply (prop_tac "rrel rv' (xf inter_t)")
+     apply (fastforce dest: hoarep_exec[OF _ _ setter_rrel, rotated])
     apply (elim impE)
-     apply (drule_tac s'=inter_t and r1="xf t'" in ccorresE_gets_the[OF cond_ccorres]; assumption?)
+     apply (frule_tac s'=inter_t and r1=rv' in ccorresE_gets_the[OF setter_ccorres]; assumption?)
        apply (fastforce intro: no_ofail)
       apply (fastforce dest: EHOther)
-     subgoal by (fastforce dest: hoarep_exec intro: setter_inv_cond)
-    apply (prop_tac "xf inter_t = xf t'")
-     apply (fastforce dest: hoarep_exec[rotated] intro: setter_xf_inv)
+     apply (intro conjI)
+           apply fastforce
+          using no_ofail apply (fastforce elim!: no_ofailD)
+         apply fastforce
+        apply (fastforce dest: hoarep_exec[OF _ _ setter_rrel, rotated])
+       apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
+      apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
+     apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
```

Comment:
I could well be wrong, but this also looks like this might be possible to compress into a single fastforce. E.g. something like 
```suggestion
     using no_ofail
     apply (fastforce elim!: no_ofailD dest: hoarep_exec[OF _ _ setter_rrel, rotated hoarep_exec[OF _ _ setter_inv_cond, rotated])
```

The advantage would be that the single fastforce is not sensitive to the order of the conjuncts should those ever change.

Reply:
I couldn't get this or something like it to work. The conjuncts are from the statement `helper` itself, so if they did change, it would be from some fairly major surgery on this rule itself. Is it worth looking into some other approach to this?

I did try getting rid of this `helper` bit, but it's useful, again because of the rather specific predicate that `exec_While_final_inv''` uses.

Reply:
It's Ok, if it doesn't work then let's leave it as is. Sometimes it's not really worth the time.